### DISCUSSION TITLE
JS 비교 연산자

### DISCUSSION THREAD

**Discussion by WooGie911** [wanted-9th-3team/frontend-study#69] [프론트 기술 질문 공간]
아래 결과를 예상해보세요.
과정에서 새롭게 알게 되었거나 인상깊은 부분에 대해 이야기 해주세요.

1. 5 > 4
2. "2" > "12"
3. "5" == 5
4. "5" === 5
5. "6" > 5
6. "apple" > "pineapple"
7. undefined == null
8. undefined === null
9. 0 == false 
10. 0 === false
11. "" == false 
12. null == false 
13. undefined == false

**Comment by michoball** [comment]
1. true
2. true
3. true
4. false
5. true
6. false
7. true
8. false
9. true
10. false
11. true
12. false
13. false

== 연산자에서는 양쪽의 타입을 맞춰서 체크한다는 사실을 알고 있어서 숫자와 문자를 비교하는 부분은 쉬웠는데 문자와 문자는 살짝 생각을 하게 했습니다. 
문자끼리는 ASCII 값을 비교한다는 사실을 생각하면 단순히 맨 첫글자부터 알파벳 순서로 비교를 하고 a 에 가까울수록 값이 작다는 사실을 알면 금방 판단할 수 있었습니다. 

뜻밖에 null 과 false, undefined 와 false , undefined와 null 을 비교하는 부분은 신선했습니다. 
특히 null 이 false와  == 로도 같지않고 undefined도 false와 ==로 같지 않다는 사실은 제가 까먹고 있었던 것 같습니다.

**Comment by WooGie911** [comment]
1. 5 > 4   => true
2. "2" > "12"   => true
3. "5" == 5   => true
4. "5" === 5   => false
5. "6" > 5   => true
6. "apple" > "pineapple"   => false
7. undefined == null   => true
8. undefined === null   => false
9. 0 == false   => true
10. 0 === false   => false
11. "" == false   => true
12. null == false   => false
13. undefined == false   => false


- 비교 연산자는 불린형을 반환한다.

- 문자열 비교는 사전편집(lexicographical)순으로 비교한다.

>  정확히는 사전순이 아니라, 유니코드 순이다. (A보다 a가 더크다)

- 비교하는 자료형이 다를 경우, 숫자형으로 바꾼다.

>  true는 1, false는 0으로 변환된 후 비교된다. ( null은 0 / undefine은 NaN으로 변환된다.)

- 동등연산자(==, equality operator)

>  0 과 false, '' 과 false을 구별하지 못한다.

>  null과 undefine은 특이케이스인데, 형변환을 하지 않고, 둘일때만 true를 반환한다.

- 일치연산자(===, strict equality operator)를 사용해야 형 변환 없이 비교한다.

- 주의할점

>  undefined나 null이 될 수 있는 변수의 경우, <, >, <=, >= 피연산자가 되지 않도록 한다.

> * 만약 쓸경우, null이나, undefined여부를 확인하는 코드를 반드시 추가한다.

>  되도록 일치연산자(===)를 사용한다.

**Comment by BaekYuri** [comment]
1. 5 > 4 -> true
2. "2" > "12" -> true
3. "5" == 5 -> true
4. "5" === 5 -> false
5. "6" > 5 -> true
6. "apple" > "pineapple" -> false
7. undefined == null -> true
8. undefined === null -> false
9. 0 == false -> true
10. 0 === false -> false
11. "" == false -> true
12. null == false -> false
13. undefined == false -> false

단순 문자열 비교는 비교연산자로도 할 수 있다는 것이 신기하다.
